### PR TITLE
Migrate label colors to Angular component

### PR DESCRIPTION
**Pull request by mi-we**

### CODE REVIEW

File: scenarioo-client/app/manage/label-colors/label-colors.component.ts [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+
+    constructor(private labelConfigurationsListResource: LabelConfigurationsListResource,
+                private labelConfigurationsResource: LabelConfigurationsResource) {
+    }
+
+    ngOnInit(): void {
+        this.loadLabelConfigurations();
+    }
+
+    onDelete(configurationToDelete) {
```
Comment (danielsuter) [inline scenarioo-client/app/manage/label-colors/label-colors.component.ts]: type parameter

File: scenarioo-client/app/manage/label-colors/label-configuration.ts [hunk 2]
```diff
@@ -0,0 +1,32 @@
+import {IFlatLabelConfiguration} from '../../generated-types/backend-types';
+
+export class LabelConfiguration implements IFlatLabelConfiguration {
+
+    backgroundColor: string;
+    foregroundColor: string;
+    name: string;
+
+    private constructor(backgroundColor: string, foregroundColor: string, name: string) {
```
Comment (danielsuter) [inline scenarioo-client/app/manage/label-colors/label-configuration.ts]: just add the public keyword in front of every parameter and you will have your public fields including the assignment

File: scenarioo-client/app/manage/label-colors/label-configuration.ts [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+        this.backgroundColor = backgroundColor;
+        this.foregroundColor = foregroundColor;
+        this.name = name;
+    }
+
+    isEmpty(): boolean {
+        return this.name === '' && this.backgroundColor === '' && this.foregroundColor === '';
+    }
+
+    isValid() {
```
Comment (danielsuter) [inline scenarioo-client/app/manage/label-colors/label-configuration.ts]: boolean

File: scenarioo-client/app/manage/label-colors/label-colors.component.ts [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+                    ...labelConfigurations.map(LabelConfiguration.fromFlatLabelConfiguration),
+                    LabelConfiguration.empty()];
+            });
+    }
+
+    private allConfigurationsAreValid() {
+        return this.labelConfigurations.every((value) => value.isValid());
+    }
+
+    private mapLabelConfigurations() {
```
Comment (danielsuter) [inline scenarioo-client/app/manage/label-colors/label-colors.component.ts]: type

File: scenarioo-client/app/manage/label-colors/label-colors.component.ts [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+    private loadLabelConfigurations() {
+        this.labelConfigurationsListResource.query()
+            .subscribe((labelConfigurations) => {
+                this.labelConfigurations = [
+                    ...labelConfigurations.map(LabelConfiguration.fromFlatLabelConfiguration),
+                    LabelConfiguration.empty()];
+            });
+    }
+
+    private allConfigurationsAreValid() {
```
Comment (danielsuter) [inline scenarioo-client/app/manage/label-colors/label-colors.component.ts]: type

File: scenarioo-client/app/manage/label-colors/label-colors.component.ts [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+
+    onDelete(configurationToDelete) {
+        this.labelConfigurations = this.labelConfigurations
+            .filter((configuration) => configuration.name !== configurationToDelete.name);
+    }
+
+    onLabelNameChanged() {
+        const lastLabelInList = this.labelConfigurations[this.labelConfigurations.length - 1];
+        if (!lastLabelInList.isEmpty()) {
+            this.labelConfigurations.push(LabelConfiguration.empty());
```
Comment (danielsuter) [inline scenarioo-client/app/manage/label-colors/label-colors.component.ts]: why do you use different styles? On line 70 you use the spread operator

File: scenarioo-client/app/manage/label-colors/label-configuration.ts [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+        return this.isEmpty() || (this.name !== '' && this.backgroundColor !== '');
+    }
+
+    static fromFlatLabelConfiguration(flatLabelConfiguration: IFlatLabelConfiguration): LabelConfiguration {
+        return new LabelConfiguration(flatLabelConfiguration.backgroundColor,
+            flatLabelConfiguration.foregroundColor,
+            flatLabelConfiguration.name);
+    }
+
+    static empty() {
```
Comment (danielsuter) [inline scenarioo-client/app/manage/label-colors/label-configuration.ts]: type

File: scenarioo-client/app/styles/scenarioo.less [hunk 8]
```diff
@@ -577,29 +577,6 @@ span.sc-text-before-label {
   margin-bottom: 1em;
 }
 
-.sc-color-field {
```
Comment (danielsuter) [inline scenarioo-client/app/styles/scenarioo.less:580]: nice!

File: scenarioo-server/build.gradle [hunk 9]
```diff
@@ -56,7 +57,7 @@ generateTypeScript {
             'org.scenarioo.rest.search.SearchEngineStatus',
             'org.scenarioo.model.docu.aggregates.usecases.UseCaseSummary',
             'org.scenarioo.model.docu.aggregates.branches.BranchBuilds',
-            'org.scenarioo.rest.sketcher.issue.dto.SketchIds',
+            'org.scenarioo.rest.sketcher.issue.dto.SketchIds'
```
Comment (danielsuter) [inline scenarioo-server/build.gradle]: why did you remove the comma? Now it shows a change where none is

File: scenarioo-server/src/main/java/org/scenarioo/rest/configuration/FlatLabelConfiguration.java [hunk 10]
```diff
@@ -0,0 +1,27 @@
+package org.scenarioo.rest.configuration;
+
+import org.scenarioo.model.configuration.LabelConfiguration;
+
+import javax.xml.bind.annotation.XmlRootElement;
+
+@XmlRootElement
```
Comment (danielsuter) [inline scenarioo-server/src/main/java/org/scenarioo/rest/configuration/FlatLabelConfiguration.java]: why do we need  the root element annotation?

**Review by danielsuter — COMMENTED** [review]
Thank you for migrating this page!

File: scenarioo-client/app/manage/label-colors/label-colors.component.ts (hunk 6 above)
Comment (mi-we) [reply scenarioo-client/app/manage/label-colors/label-colors.component.ts]: Yep, that didn't catch my attention. I will use the spread operator here as well.

File: scenarioo-client/app/manage/label-colors/label-configuration.ts (hunk 2 above)
Comment (mi-we) [reply scenarioo-client/app/manage/label-colors/label-configuration.ts]: Thanks for the tip! Will do that

File: scenarioo-server/build.gradle (hunk 9 above)
Comment (mi-we) [reply scenarioo-server/build.gradle]: Sorry, was adding the new DTO at the end of the list first and must have overlooked that there was already a comma before.

File: scenarioo-server/src/main/java/org/scenarioo/rest/configuration/FlatLabelConfiguration.java (hunk 10 above)
Comment (mi-we) [reply scenarioo-server/src/main/java/org/scenarioo/rest/configuration/FlatLabelConfiguration.java]: It does not seem to be necessary. I removed it.

**Comment by mi-we** [comment]
@danielsuter Thanks for your review. I addressed all the points. Can you take another look if everything is ok from your side? Thanks!

**Comment by bruderol** [comment]
@mi-we es gibt da konflikte sonst könnte man das ev mal mergen ???

**Comment by bruderol** [comment]
@mi-we I think this makes PR #800 obsolete ?? did you consider what they did there?

Can we then close this PR #800 and delete the other branch?

**Comment by mi-we** [comment]
> @mi-we I think this makes PR #800 obsolete ?? did you consider what they did there?
> 
> Can we then close this PR #800 and delete the other branch?

Hi @bruderol 

When we decided on working on the label colors component we agreed on redoing the changes on this branch. So yes, we could remove the other changes now and merge this one, that is, if the builds run through now after I resolved the conflicts.

**Comment by bruderol** [comment]
Thanks  @mi-we I also closed the other PR now - and I know, that we decided this - my question was just whether it makes still sense for some reason to keep it open ??

Once a PR does not have any valuable content anymore, we should close them
